DragaliaAPI/Controllers/Dragalia/DragonController.cs

```diff
@@ -0,0 +1,510 @@
+            this.DeviceAccountId
+        );
+
+        await unitRepository.SaveChangesAsync();
+
+        return Ok(
+            new DragonBuildupData(
+                updateDataList,
+                new DeleteDataList(
+                    request.grow_material_list
+                        .Where(x => x.type == EntityTypes.Dragon)
+                        .Select(x => new AtgenDeleteDragonList() { dragon_key_id = (ulong)x.id }),
+                    new List<AtgenDeleteTalismanList>(),
+                    new List<AtgenDeleteWeaponList>(),
+                    new List<AtgenDeleteAmuletList>()
+                ),
+                new()
+            )
+        );
+    }
+
+    private void DragonLevelUp(
+        IEnumerable<GrowMaterialList> materials,
+        ref DbPlayerDragonData playerDragonData,
+        ref Dictionary<int, int> usedMaterials
+    )
+    {
+        //TODO: For now we'll trust the client to not allow leveling up/enhancing beyond allowed limits
+        byte maxLevel = DragonConstants.GetMaxLevelFor(
+            _dragonDataService.GetData((int)playerDragonData.DragonId).Rarity,
+            playerDragonData.LimitBreakCount
+        );
+        int gainedHpAugs = 0;
+        int gainedAtkAugs = 0;
+        foreach (GrowMaterialList materialList in materials)
+        {
+            if (materialList.type == EntityTypes.Material)
+            {
+                switch ((Materials)materialList.id)
+                {
+                    case Materials.Dragonfruit:
+                    case Materials.RipeDragonfruit:
+                    case Materials.SucculentDragonfruit:
+                        playerDragonData.Exp = Math.Min(
+                            playerDragonData.Exp
+                                + (
+                                    UpgradeMaterials.buildupXpValues[(Materials)materialList.id]
+                                    * materialList.quantity
+                                ),
+                            DragonConstants.XpLimits[maxLevel - 1]
+                        );
+                        break;
+                    case Materials.AmplifyingDragonscale:
+                        playerDragonData.AttackPlusCount = (byte)
+                            Math.Min(
+                                playerDragonData.AttackPlusCount + materialList.quantity,
+                                DragonConstants.MaxAtkEnhance
+                            );
+                        break;
+                    case Materials.FortifyingDragonscale:
+                        playerDragonData.HpPlusCount = (byte)
+                            Math.Min(
+                                playerDragonData.HpPlusCount + materialList.quantity,
+                                DragonConstants.MaxHpEnhance
+                            );
+                        break;
+                    default:
+                        throw new ArgumentException("Invalid MaterialList");
```

Comment:
Doesn't this duplicate material validation logic in the first part of the controller? My preference would be to keep this one if so